ENH: general concat with ExtensionArrays through find_common_type

jorisvandenbossche opened the pull request:
Exploring what is being discussed in https://github.com/pandas-dev/pandas/issues/22994. @TomAugspurger your idea seems to be working nicely! (it almost removes as much code than it adds (apart from tests/docs), ànd fixes the EA concat bugs ;))

Few notes compared to proposal in #22994:

- Since we already have a `find_common_type` function, decided to use this as the "get_concat_dtype", since it seems this does what is needed for concat
- Extending the EA interface with a `ExensionDtype._get_common_type` method that is used in pandas' `find_common_type` function to dispatch the logic to the extension type

What I already handled:

- general protocol, using this for concat with axis=0 (eg concatting series) and using it as example for IntegerDtype
- handling categoricals this way (which removes the `concat_categorical` helper). This turned up a few cases where we have value-dependent behaviour right now, which we can't easily preserve (mainly regarding NaNs and int dtype, see below)

Still need to handle sparse (those has failing tests now) and maybe datetime, and check other failures.

jorisvandenbossche committed:
initial find_common_type/_get_common_type + tests for IntegerDtype

jorisvandenbossche committed:
handle categoricals

jorisvandenbossche commented on pandas/core/arrays/integer.py:
```diff
@@ -95,6 +95,15 @@ def construct_array_type(cls) -> Type["IntegerArray"]:
+        np_dtype = np.find_common_type([t.numpy_dtype for t in dtypes], [])
+        if np.issubdtype(np_dtype, np.integer):
+            return _dtypes[str(np_dtype)]
+        return None
```
For now, this only handles concatting of several EA IntegerDtypes. Of course, we can expand this to also handle mix of EA Int dtype with numpy int dtypes, and also handle mix of integer dtype with non-integer (like coercing bool to integer, coercing to floats, ...). 
But this could also be done separately in a follow-up  to fully flesh this out, once the protocol is in place (and at least the above implements the minimal to show it works).

TomAugspurger replied:
Agreed with keeping it simple for now. We can try to achieve the behavior on master (aside from fixing a few bugs) and then discuss whether we should handle casting.

jbrockmendel replied:
agreed

jreback replied:
sgtm

jorisvandenbossche commented on pandas/core/dtypes/concat.py:
```diff
@@ -99,14 +103,36 @@ def is_nonempty(x) -> bool:
+            target_dtype = find_common_type([x.dtype for x in to_concat])
+
+            def cast(arr, dtype):
+                if (
+                    is_categorical_dtype(arr.dtype)
+                    and isinstance(dtype, np.dtype)
+                    and np.issubdtype(dtype, np.integer)
+                ):
+                    # problem case: categorical of int -> gives int as result dtype,
+                    # but categorical can contain NAs -> fall back to object dtype
+                    try:
+                        return arr.astype(dtype, copy=False)
+                    except ValueError:
+                        return arr.astype(object, copy=False)
```
This complication is to try to preserve some of the value-dependent behaviour of Categorical (in case of integer categories: missing values present or not?)

Eg when concatting integer categorical with integer series:

```
pd.concat([pd.Series([1, 2], dtype="category"), pd.Series([3, 4])]) 
-> results in int dtype
pd.concat([pd.Series([1, None], dtype="category"), pd.Series([3, 4])]) 
-> results in object dtype
```

Currently, when concatting, a Categorical with integer categories gets converted to int numpy array if no missing values are present, but object numpy array if missing values are present (to preserve the integers)

jorisvandenbossche commented:
I needed to make one change to the tests related to categorical (another value-dependent behaviour). Assume those examples involving integer categorical with float arrays:

```
In [13]: pd.concat([pd.Series([1, None], dtype="category"), pd.Series([3.5, 4.5])])    
Out[13]: 
0      1
1    NaN
0    3.5
1    4.5
dtype: object

In [14]: pd.concat([pd.Series([1, 2], dtype="category"), pd.Series([3.5, 4.5])])    
Out[14]: 
0    1.0
1    2.0
0    3.5
1    4.5
dtype: float64
```

With a `find_common_type`, we will always find float. But the old behaviour is that the integer categorical becomes object if there are missing values present. 
But I would say: if you are concatting with a float anyway, it doesn't make sense to try to provide the "integer categories" by keeping it object instead of casting to float.

TomAugspurger commented on pandas/core/dtypes/base.py:
```diff
@@ -322,3 +323,33 @@ def _is_boolean(self) -> bool:
         """
         return False
+
+    def _get_common_type(self, dtypes: List[DtypeObj]) -> Optional[DtypeObj]:
```
Mmm can we keep the return type as `ExtensionDtype`? Do you envision cases where we'd like to return a plain NumPy dtype?

Oh... I suppose tz-naive DatetimeArray might break this, since it wants to return a NumPy dtype...

jorisvandenbossche replied:
Yes, that was my first thought as well. But, right now, eg Categorical can end up with any kind of numpy dtype (depending on the dtype of its categories).

As long as not yet all dtypes have a EA version, I don't think it is feasible to require ExtensionDtype here

jorisvandenbossche committed:
handle sparse

jorisvandenbossche committed:
handle non-initialized CategoricalDtype in find_common_type

jorisvandenbossche commented:
I ran into one other hairy question:

- Should `_get_common_type` be able to handle "not fuly initialized" dtype objects? Like a CategoricalDtype without categories? (which we have to handle "category" string).  
  And if so, what is the expected behaviour?


Right now I added some special cases in `Categorical._get_common_type` to have the find_common_type tests passing, but not sure it is necessarily doing the correct thing in a general way.

jorisvandenbossche committed:
handle datetimelike special case

jbrockmendel commented on pandas/core/arrays/integer.py:
```diff
@@ -95,6 +95,15 @@ def construct_array_type(cls) -> Type["IntegerArray"]:
         """
         return IntegerArray
 
+    def _get_common_type(self, dtypes: List[DtypeObj]) -> Optional[DtypeObj]:
```
should this be common_type or common_dtype?  we've been loose about this distinction so far and i think it has caused amibiguity

jorisvandenbossche replied:
I don't care that much. I mainly used "type", because it is meant to be used in `find_common_type`.

(that `find_common_type` name is inspired on the numpy function, and that one actually handles both dtypes and scalar types, which I assume is the reason for the name. The pandas version, though, doesn't  really make the distinction, so could have been named "find_common_dtype")

jorisvandenbossche replied:
Renamed to "common_dtype" instead of "common_type". The internal function that uses this is still `find_common_type`, but that name from numpy is actually a misnomer here, since we are only dealing with dtypes, and not scalar types.

jbrockmendel replied:
Thanks for indulging me on this nitpick

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -2349,9 +2349,9 @@ def _can_hold_na(self):
     @classmethod
     def _concat_same_type(self, to_concat):
-        from pandas.core.dtypes.concat import concat_categorical
+        from pandas.core.dtypes.concat import union_categoricals
```
is this changed on purpose? it’s an api change

jorisvandenbossche replied:
This is inside `_concat_same_type`, so no external API change 
(and `_concat_same_type` always was intended to only handle same-dtype arrays)

jbrockmendel commented on pandas/core/internals/concat.py:
```diff
@@ -65,13 +66,15 @@ def concatenate_block_managers(
 
-            if not blk.is_extension or blk.is_datetimetz or blk.is_categorical:
+            if not blk.is_extension or blk.is_datetimetz:
                 # datetimetz and categorical can have the same type but multiple
```
the comment here may need to be updated

jorisvandenbossche replied:
Removed the comment (and the datetimetz special case) entirely now

jreback reviewed (changes requested):
generally good concept will look in more detail later

jbrockmendel commented on pandas/core/dtypes/base.py:
```diff
@@ -322,3 +323,34 @@ def _is_boolean(self) -> bool:
+
+        Returns
+        -------
+        Common dtype (np.dtype or ExtensionDtype) or None
```
what does None mean here?

jorisvandenbossche replied:
See the docstring: "If no common dtype exists, return None". 
I can certainly try to expand on what this exactly means, if you have suggestions to make it clearer

jbrockmendel replied:
i guess im looking for something about what are the consequences if None is returned.  I think the answer is "we try the next EADtype in `find_common_type`", but could also imagine "we end up with object dtype" or "we cast everything to numpy and get whatever `np.concatenate` gives back"

jorisvandenbossche replied:
Again, see the docstring, it is saying those things, but probably not in enough words (will try to expand that). Can you try to be more precise about what is not clear in the text above?

jbrockmendel replied:
yep, its pretty clear now that ive got some caffeine in me, thanks

jorisvandenbossche replied:
Expanded the docstring a bit anyway

jbrockmendel commented on pandas/core/dtypes/cast.py:
```diff
@@ -1474,7 +1474,12 @@ def find_common_type(types):
+                res = t._get_common_type(types)
+                if res is not None:
+                    return res
+        return np.dtype("object")
```
i like having this logic live here.

@TomAugspurger the "negotiation protocol" would affect what order we iterate over types?

TomAugspurger replied:
Dunno, but I doubt it. At the end of the day we have established behavior that order matters (I think).

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -58,6 +64,40 @@ def get_dtype_kinds(l):
 
+def _cast_to_common_type(arr: ArrayLike, dtype: DtypeObj) -> ArrayLike:
+    """
+    Helper function for `arr.astype(common_type)` but handling all special
```
common_type or common_dtype?

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -58,6 +64,40 @@ def get_dtype_kinds(l):
+        and arr.dtype.kind in ["m", "M"]
+        and dtype is np.dtype("object")
+    ):
+        # wrap datetime-likes in EA to ensure astype(object) gives Timestamp/Timedelta
```
+1 quality comment

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -93,28 +133,25 @@ def is_nonempty(x) -> bool:
-        # this must be prior to concat_datetime,
-        # to support Categorical + datetime-like
-        return concat_categorical(to_concat, axis=axis)
+        if isinstance(to_concat[0], ExtensionArray):
```
need to be a little careful: `is_extension_array_dtype(tda) != isinstance(tda, ExtensionArray)`

jbrockmendel replied:
(not sure if it matters in this case, but i saw checks of both varieties, so worth mentioning)

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -93,28 +133,25 @@ def is_nonempty(x) -> bool:
 
-    elif _contains_datetime or "timedelta" in typs or _contains_period:
+    elif _contains_datetime or "timedelta" in typs:
         return concat_datetime(to_concat, axis=axis, typs=typs)
```
if we do the DTA/TDA casting above, and do `isinstance(obj, ExtensionArray)` checks, can all of the dt64/td64 cases be handled by the EA code above?

jorisvandenbossche replied:
I don't think so, because they are not using ExtensionDtype.

jbrockmendel commented:
I like this idea, will want to give it another look after more caffeine.  Haven't looked at the tests or Categorical-specific nuances yet.

jbrockmendel commented on pandas/core/arrays/categorical.py:
```diff
@@ -2349,9 +2349,9 @@ def _can_hold_na(self):
+        from pandas.core.dtypes.concat import union_categoricals
 
-        return concat_categorical(to_concat)
+        return union_categoricals(to_concat)
```
IIRC only a relatively small part of the logic of concat_categorical/union_categoricals is actually needed here.  I'd prefer for that to live here and for union_categoricals to call it, rather than the other way around (since union_categoricals handles a lot of cases).  Could be considered orthogonally to this PR.

jorisvandenbossche replied:
>  I'd prefer for that to live here and for union_categoricals to call it, rather than the other way around

Yes, that's indeed a good idea (union_categoricals does way much more that is all not needed here)

jbrockmendel replied:
are you planning to update this, or is that a topic for a separate PR?

jorisvandenbossche committed:
update docstring and comment

jorisvandenbossche committed:
ignore mypy

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -95,6 +95,17 @@ def construct_array_type(cls) -> Type["IntegerArray"]:
+        # for now only handle other integer types
+        if not all(isinstance(t, _IntegerDtype) for t in dtypes):
+            return None
+        np_dtype = np.find_common_type(
```
is there a reason you are not using pandas `pandas.core.dtypes.cast.find_common_type` ?

jorisvandenbossche replied:
Because this are all numpy integer dtypes, so there is no need to use our version that can handle our custom logic/dtypes

jreback commented on pandas/core/arrays/integer.py:
```diff
@@ -95,6 +95,17 @@ def construct_array_type(cls) -> Type["IntegerArray"]:
         """
         return IntegerArray
 
+    def _get_common_type(self, dtypes: List[DtypeObj]) -> Optional[DtypeObj]:
```
why is this a private method on the Dtype? get_common_type (or get_common_dtype) seems fine

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -58,6 +64,40 @@ def get_dtype_kinds(l):
     return typs
 
 
+def _cast_to_common_type(arr: ArrayLike, dtype: DtypeObj) -> ArrayLike:
```
why is this private? everything else is not, so this seems odd

jorisvandenbossche replied:
This is an internal helper function for this file, which we indicate by using a leading underscore (the other functions here are used elsewhere in the pandas codebase)

jreback commented on pandas/tests/arrays/integer/test_concat.py:
```diff
@@ -0,0 +1,26 @@
+@pytest.mark.parametrize(
+    "to_concat_dtypes, result_dtype",
+    [
+        (["Int64", "Int64"], "Int64"),
```
what about nullable int with numpy int -> object?

jorisvandenbossche replied:
I think it can probably return nullable int instead of object dtype, but see https://github.com/pandas-dev/pandas/pull/33607/files#r410127707, fleshing out the exact behaviour for concat with nullable integers is for a follow-up

jreback commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -610,11 +610,11 @@ def test_concat_categorical_3elem_coercion(self):
         s3 = pd.Series([1, 2, 1, 2, np.nan])
 
-        exp = pd.Series([1, 2, np.nan, 2, 1, 2, 1, 2, 1, 2, np.nan], dtype="object")
+        exp = pd.Series([1, 2, np.nan, 2, 1, 2, 1, 2, 1, 2, np.nan], dtype="float")
```
this would be an api change can you add a release note

jorisvandenbossche replied:
Yes, but can you check https://github.com/pandas-dev/pandas/pull/33607#issuecomment-615168747 and indicate if we are fine with that change?

jorisvandenbossche commented:
> [@jreback] why is this a private method on the Dtype? get_common_type (or get_common_dtype) seems fine

This was briefly touched upon in https://github.com/pandas-dev/pandas/issues/22994#issuecomment-614249196. The reason to make this private, is because the method is not meant to be used by *end users*, but only meant to be implemented by *EA authors*. And thus, we think that a private method is appropriate here (similarly as `EDtype._is_boolean/_is_numeric`, `EA._from_sequence`, `EA._concat_same_type`, ... are also all private methods to the user, but part of the public EA interface).

But what actually might be useful to publicly expose (eg in `pd.api.types`) is the `find_common_type` function (which this `_get_common_type` is helping implement). I can image this could be useful for eg dask? (@TomAugspurger ?)

TomAugspurger commented:
A public `find_common_type` might be useful, but don't let Dask be the motivator. We would get the result dtypes by doing an actual concat on two empty DataFrames.

jorisvandenbossche commented:
More comments here?

Note I added some questions regarding the `ExtensionDtype._get_comon_type` interface at the base implementation:

- do we guarantee that `dtypes` is already deduplicated? (passing a list of unique dtypes)
- do we call this method if `len(dtypes) == 1`, or should this method need to handle that case?
- does this method need to handle "non-fully-initialized" dtypes? (for concat, you will never run into this, since the dtypes are always coming from actual arrays)

jorisvandenbossche commented:
> do we guarantee that dtypes is already deduplicated? (passing a list of unique dtypes)

Currently I didn't do this yet in `find_common_type`, but probably should. 
@TomAugspurger you wrote this part in your pseudo-code in the issue as `{x.dtype for x in arrays}`, but I don't think that sets preserve order? (which we want here, possible workaround seems to be `dict.fromkeys`, this has much less overhead as pd.unique)

TomAugspurger commented:
I don't thin sets are required to preserve order.

You could maintain a mapping like (pseducode)

```python
seen = set()
uniques = {x.dtype for x in arrays}
for array in arrays:
    if array.dtype not in seen:
        common_dtype = array.dtype._get_common_type(uniques)
    seen.add(array.dtype)
    if common_dtype:
        break
```

Will try to look again in a bit, but don't wait up for me.

jbrockmendel commented:
API Questions:

- Should `_get_common_type` be a classmethod?
- Are we implicitly assuming that `self.dtype` is a member of `dtypes`?
    - If so, can we make that explicit
- In the case (mentioned in a comment on the base class) where `dtypes` is a singleton, can we require that the return value be that dtype?
- Can we nail down the common_type vs common_dtype semantics before pushing this through?  I feel like that has been a major cause of trouble with the existing concat code.

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -58,6 +64,40 @@ def get_dtype_kinds(l):
+    """
+    Helper function for `arr.astype(common_dtype)` but handling all special
+    cases.
+    """
```
do you have a plan in mind to make this function unnecessary?  it is special-casing pandas-internal EAs which i think we want to avoid

(Not a blocker for this PR because AFAICT this is necessary and we do this in the status quo anyway, just curious)

jorisvandenbossche replied:
Yeah, so this is indeed only needed to preserve value-based specific behaviour, not because I like this :)

Ideally we get rid of this, but I am not sure it is entirely possible with the numpy integer dtypes, since it will always depend on the presence of NaNs whether it can preserve int or not. The way to go here is probably the nullable integer dtypes to have consistent behaviour.

jorisvandenbossche commented:
> Are we implicitly assuming that `self.dtype` is a member of `dtypes`? 

Right now, this is the case yes. What do you mean with making that explicit? (just better documenting it?)

> Should `_get_common_type` be a classmethod?

It could be, but personally I don't really see to not make it an instance method. Yes, `self` will always be in the list of `dtypes` (see question above), but I can assume that for the implementation it can be useful to know self (instead of needing to infer it from the list of dtypes). Not that this should impact the *result*, I only think it can be handy in the implementation (but this is rather hypothetical feeling at the moment). 
The main thing is: we always call this from the instance, so I don't see much value in making it a class method instead of instance method.

> In the case (mentioned in a comment on the base class) where `dtypes` is a singleton, can we require that the return value be that dtype?

This is the only sensible return value in such a case (an EA doing anything different will behave very strange). But what do you mean with "require"? Better document it? Or we could have a base extension test that asserts it? 

But see also my question above (https://github.com/pandas-dev/pandas/pull/33607#issuecomment-617403497): in principle we don't even need to call this `_get_common_type` method if there is only one dtype in the list. In the concat code / `find_common_type` code, we can simply use this dtype and shortcut calling `_get_common_type`.

> Can we nail down the common_type vs common_dtype semantics before pushing this through? I feel like that has been a major cause of trouble with the existing concat code.

Can you explain this a bit more? I am not aware of any troubles regarding that.

jorisvandenbossche commented:
I have also been reading up on numpy dtype proposals regarding this. Specifically this part of the draft NEP relates to "common dtypes": https://github.com/numpy/numpy/blob/f07e25cdff3967a19c4cc45c6e1a94a38f53cee3/doc/neps/nep-0042-new-dtypes.rst#common-dtype-operations

It's actually quite similar in idea, but some notable differences:

- The dtype special method is limited to a *single* other dtype (they have `Dtype.__common_dtype__(self, other)` where `other` is a single dtype), instead of passing the list of all dtypes
- They split the "common dtype" special method in two: determine the common *class* (from 2 different dtype classes), and determine the common *instance* (from 2 instances of the *same* class). And it is actually split in 3, because those 2 steps are then combined with as cast-lookup.

The first is an interesting idea. The second might only complicate it more for *current* pandas (although I understand its desire to split it up on steps to avoid duplicate logic in different parts). 
And I also don't know if first doing it on the *class* level will work for pandas' current promotion logic. Eg the common dtype of Categorical + int dtype depends on the dtype of the categories of the Categorical.

jbrockmendel commented:
>> Can we nail down the common_type vs common_dtype semantics before pushing this through? I feel like that has been a major cause of trouble with the existing concat code.

> Can you explain this a bit more? I am not aware of any troubles regarding that.

"trouble" in that we haven't been consistent about where some logic belongs.  e.g. DTA._concat_same_type requires a single dtype, not just same-type.

jorisvandenbossche commented:
With this PR, `EA._concat_same_type` will only be called with arrays of all the same dtype. 
(and which needs to be better documented now, as that was not clear from its docstring what the expectations were before)

Does that answer your concern? Otherwise you will still need to clarify further what you mean.

jbrockmendel commented:
> Does that answer your concern? Otherwise you will still need to clarify further what you mean.

Close enough.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into concat-protocol

jorisvandenbossche committed:
common_type -> commong_dtype

jorisvandenbossche committed:
ensure deduplicated list of dtypes is passed

jorisvandenbossche committed:
add very basic base extension test

jorisvandenbossche committed:
document API change

jorisvandenbossche committed:
update EA interface docs

jorisvandenbossche committed:
add type annotation on find_common_type

jorisvandenbossche commented:
I made a bunch of updates (renamed to use "dtype" instead of "type", documented the public API change and updated the EA interface docs, added a base extension test). 
Ready for another review (or approval ;)) round

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1450,8 +1450,16 @@ def find_common_type(types):
     if any(isinstance(t, ExtensionDtype) for t in types):
-        return np.object
+        for t in types:
+            if isinstance(t, ExtensionDtype):
```
this seems a bit arbitrary. we try the types in order to see if they are a unifying type? do we actually test this?

jorisvandenbossche replied:
Yes, so it is not arbitrary, but in order. 
(eg the order of the arguments passed to concat. This order for example also determines the class used for the return value)

I don't think we have internally dtypes where this order actually matters, right now (meaning, where a different order would give a different result). So we don't *really* test the order-dependency.

jorisvandenbossche replied:
Tom and I also had some discussion about this in https://github.com/pandas-dev/pandas/issues/22994#issuecomment-614203960

jreback commented:
thanks @jorisvandenbossche very nice

jorisvandenbossche commented:
And thanks all for the review!
